nixos/nextcloud: add extraSecrets option

kraftnix opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->
Add the `extraSecrets` option to allow setting secret values in `config.ini` seamlessly using the existing `nix_read_secret` and LoadCredential integrations for handling secrets in `config.ini`.

Currently `nix_read_secret` is only used for setting hardcoded secrets, but it is really useful to expose this to allow user settable secrets without needing to use the current `secretFile` option which requires using an entire secret json file (not very amenable to config merging).

I currently use this to easily set the shared oidc secret location for the oidc_login plugin, but it can be reused for any secret value.
```nix
oidc_login_client_secret = "/run/secrets/nextcloud_oidc_secret" 
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Ma27 reviewed (changes requested):
I'm not sure I follow. Which `config.ini`? Do you mean `config.php`?

Also, how is this different from the `extraSecrets` option?

kraftnix committed:
nixos/nextcloud: add extraSecrets option

Uses existing `nix_read_secret` and LoadCredential to read contents of a
file into an entry in `config.php`

Ma27 commented:
> All it does is expose already existing, already implemented functionality to the user for adding their own secrets to the Nextcloud config without constantly rewriting a single secret JSON file

Which means another thing we'd potentially need a migration path for if we need to change something around here and the downside still stands, confusing people on what to do.

I'm curious though what the other maintainers think (cc @provokateurin @dotlambda @britter and the author of the current secret handling @networkException ).

provokateurin commented:
I'm not against this, however I'm not sure how the example you gave works. I also use `user_oidc`, but I don't see a way to configure the secret like you did (except for some custom PHP code in one of your config.php files :thinking:
Besides this example, how many secrets are actually stored in config.php? Usually they are supposed to be stored encrypted (with the instance secret) in the DB, so there shouldn't be many use-cases for this.

provokateurin reviewed (changes requested):
Could you please add a test for this? You could just add some secret and read it back via `nextcloud-occ config:system:get <key>`

Ma27 commented on nixos/modules/services/web-apps/nextcloud.nix:
```diff
@@ -383,6 +391,19 @@ in
       '';
       example = "/mnt/nextcloud-file";
     };
+    extraSecrets = lib.mkOption {
```
The other module I'm aware of that does this is forgejo and it calls this option `secrets`. Should we do the same?
Kinda unfortunate that this isn't covered in rfc42.

provokateurin replied:
:+1: from me, the `extra` part doesn't help much IMO.
